File: kotlin/sdk/src/commonMain/kotlin/com/basecamp/fizzy/resilience/ResilienceConfig.kt
package com.basecamp.fizzy.resilience

import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds

/**
 * Configuration for all resilience components.
 *
 * Provides sensible defaults for circuit breaker, bulkhead, and rate limiter.
 * Override individual settings via the builder DSL.
 */
data class ResilienceConfig(
    val circuitBreaker: CircuitBreakerConfig = CircuitBreakerConfig(),
    val bulkhead: BulkheadConfig = BulkheadConfig(),
    val rateLimiter: RateLimiterConfig = RateLimiterConfig(),
)

data class CircuitBreakerConfig(
    /** Number of failures before opening the circuit. */
    val failureThreshold: Int = 5,
    /** Duration the circuit stays open before transitioning to half-open. */
    val resetTimeout: Duration = 30.seconds,
    /** Number of successful calls in half-open to close the circuit. */
    val halfOpenSuccessThreshold: Int = 2,
    /** Status codes considered failures. */
    val failureStatusCodes: Set<Int> = setOf(500, 502, 503, 504),
)

data class BulkheadConfig(
    /** Maximum concurrent requests. */
    val maxConcurrent: Int = 25,
    /** Maximum queue depth when at capacity. */
    val maxQueue: Int = 50,
    /** How long a queued request waits before timing out. */
    val queueTimeout: Duration = 10.seconds,
)

data class RateLimiterConfig(
    /** Maximum requests per window. */
    val maxRequests: Int = 50,
    /** Time window for the rate limit. */
    val window: Duration = 1.seconds,
    /** Minimum time between consecutive requests. */
    val minInterval: Duration = 20.milliseconds,
)

File: swift/Sources/Fizzy/Resilience/ResilienceConfig.swift
import Foundation

/// Configuration for the resilience layer.
///
/// Controls circuit breaker, bulkhead, and rate limiter behavior.
/// All values have sensible defaults.
public struct ResilienceConfig: Sendable {
    /// Circuit breaker configuration.
    public let circuitBreaker: CircuitBreakerConfig

    /// Bulkhead configuration.
    public let bulkhead: BulkheadConfig

    /// Rate limiter configuration.
    public let rateLimiter: RateLimiterConfig

    public init(
        circuitBreaker: CircuitBreakerConfig = CircuitBreakerConfig(),
        bulkhead: BulkheadConfig = BulkheadConfig(),
        rateLimiter: RateLimiterConfig = RateLimiterConfig()
    ) {
        self.circuitBreaker = circuitBreaker
        self.bulkhead = bulkhead
        self.rateLimiter = rateLimiter
    }

    /// Default resilience configuration.
    public static let `default` = ResilienceConfig()
}

/// Circuit breaker configuration.
public struct CircuitBreakerConfig: Sendable {
    /// Number of consecutive failures before opening the circuit.
    public let failureThreshold: Int
    /// Duration the circuit stays open before entering half-open state.
    public let resetTimeout: TimeInterval
    /// Number of successful probes in half-open state before closing.
    public let successThreshold: Int

    public init(failureThreshold: Int = 5, resetTimeout: TimeInterval = 30, successThreshold: Int = 2) {
        self.failureThreshold = failureThreshold
        self.resetTimeout = resetTimeout
        self.successThreshold = successThreshold
    }
}

/// Bulkhead configuration.
public struct BulkheadConfig: Sendable {
    /// Maximum number of concurrent requests.
    public let maxConcurrent: Int

    public init(maxConcurrent: Int = 10) {
        self.maxConcurrent = maxConcurrent
    }
}

/// Rate limiter configuration.
public struct RateLimiterConfig: Sendable {
    /// Maximum number of requests per window.
    public let maxRequests: Int
    /// Window duration in seconds.
    public let windowSeconds: TimeInterval

    public init(maxRequests: Int = 50, windowSeconds: TimeInterval = 10) {
        self.maxRequests = maxRequests
        self.windowSeconds = windowSeconds
    }
}
